### PR TITLE
Safronov Mikhail. Lab2. Var 1.

### PR DESCRIPTION
**Pull request by yourd8ne**
An implementation of LLVM Pass, which instruments the functions in a module by adding calls to `instrument_start()` at the beginning and `instrument_end()` at the end of each function.
The test from which the ir was made
```c++
// test.cpp

void empty() {
    return;
}

int mult(int a) {
    return a * a;
}

int increment(int x) {
    return x + 1;
}

int complexFunction(int a, int b) {
    int c = a + b;
    c = c * 2;
    return c;
}
```

### CODE REVIEW

File: llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    llvm::Function *func_f_f =
+        llvm::dyn_cast<llvm::Function>(instrStartFunc.getCallee());
+    llvm::Function *func_l_f =
+        llvm::dyn_cast<llvm::Function>(instEndFunc.getCallee());
+
+    // Проверка, добавлен ли уже вызов instrument_start
+    for (auto *U : func_f_f->users()) {
+      if (auto *callInst = llvm::dyn_cast<llvm::CallInst>(U)) {
+        if (callInst->getFunction() == &func) {
+          startInserted = true;
```
Comment (aobolensk) [inline llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp:33]:
```suggestion
          startInserted = true;
          break;
```

File: llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        }
+      }
+    }
+
+    // Проверка, добавлен ли уже вызов instrument_end
+    for (auto *U : func_l_f->users()) {
+      if (auto *callInst = llvm::dyn_cast<llvm::CallInst>(U)) {
+        if (callInst->getFunction() == &func) {
+          endInserted = true;
```
Comment (aobolensk) [inline llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp:43]:
```suggestion
          endInserted = true;
          break;
```

File: llvm/test/lab2/safronov_mikhail/instrumented_test_functions.ll [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+define dso_local void @instrument_start() {
+entry:
+  ret void
+}
+
+define dso_local void @instrument_end() {
+entry:
+  ret void
+}
```
Comment (aobolensk) [inline llvm/test/lab2/safronov_mikhail/instrumented_test_functions.ll:11]: Add a test case where these functions do not exist in .ll file

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+//#include "llvm/IR/IRBuilder.h"
```
Comment (aobolensk) [inline llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp]:
```suggestion
```

File: llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      }
+    }
+
+    return llvm::PreservedAnalyses::all();
+  }
+};
+
+extern "C" LLVM_ATTRIBUTE_WEAK ::llvm::PassPluginLibraryInfo
+llvmGetPassPluginInfo() {
+  llvm::errs() << "Loading InstrumentFunctions Plugin\n"; // Отладочный вывод
```
Comment (aobolensk) [inline llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp]:
```suggestion
```

File: llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  return {
+      LLVM_PLUGIN_API_VERSION, "instr_func", "0.1", [](llvm::PassBuilder &PB) {
+        PB.registerPipelineParsingCallback(
+            [](llvm::StringRef name, llvm::FunctionPassManager &FPM,
+               llvm::ArrayRef<llvm::PassBuilder::PipelineElement>) -> bool {
+              if (name == "instr_func") {
+                llvm::errs()
+                    << "Registering InstrumentFunctions Pass\n"; // Отладочный
+                                                                 // вывод
```
Comment (aobolensk) [inline llvm/labs/lab2/safronov_mikhail/Instrumentation_func.cpp]:
```suggestion
```

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: llvm/test/lab2/safronov_mikhail/instrumented_test_functions.ll (hunk 3 above)
Comment (yourd8ne) [reply llvm/test/lab2/safronov_mikhail/instrumented_test_functions.ll:11]: struct where these functions do not exist in .ll file

Comment (m-ly4) [reply llvm/test/lab2/safronov_mikhail/instrumented_test_functions.ll:11]: There is required are separate test file without definitions of `instrument_start`/`instrument_end` functions to check that their declarations are created in IR on insert of the calls.

File: llvm/test/lab2/safronov_mikhail/instrumented_test_functions.ll [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+; CHECK-LABEL: @_Z5emptyv
+; CHECK: call void @instrument_start()
+; CHECK-NEXT: call void @instrument_end()
+; CHECK-NEXT: ret void
+
+define dso_local noundef i32 @_Z4multi(i32 noundef %a) #0 {
+entry:
```
Comment (m-ly4) [inline llvm/test/lab2/safronov_mikhail/instrumented_test_functions.ll:38]: Could you please also add a test case with a function that already has calls to `instrument_start`/`instrument_end` functions.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by aobolensk — APPROVED** [review]

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**